ir/solver.h

```diff
@@ -15,18 +15,11 @@
 // TODO: This should implement AbstractRepCheckedNode<Constraint>.
 using Constraint = IR::Expression;
 
-/// Comparator to compare SymbolicVariable pointers.
-struct SymbolicVarComp {
-    bool operator()(const IR::SymbolicVariable *s1, const IR::SymbolicVariable *s2) const {
-        return s1->operator<(*s2);
-    }
-};
-
 /// This type maps symbolic variables to their value assigned by the solver.
-using SymbolicMapping = boost::container::flat_map<const IR::SymbolicVariable *,
```

Comment:
Note that abseil flat hash_map / hash_set are different from Boost ones. The latter are essentially sorted vectors, so here `sizeof(SymbolicMapping)` could be increased. Not sure if this something that does matter.

Reply:
In this case it should not make much of a difference. We introduced them to save some memory usage for larger programs but this has since been fixed iirc. The most important part is that the hashing works correctly.

Unfortunately, the impact of these maps is hard to evaluate because I do not have access to the largest programs any more. I am seeing some small regression in performance running `./backends/p4tools/modules/testgen/testgen-gtest --gtest_filter="*Benchmark*"`. `35s` vs `31s`.  It might be because the hash function implementation is flawed.

Reply:
No, that is a regression that I have not pushed to a PR yet. Please apply this patch: 
```
diff --git a/backends/p4tools/modules/testgen/testgen.cpp b/backends/p4tools/modules/testgen/testgen.cpp
index e825727bf..52904cf09 100644
--- a/backends/p4tools/modules/testgen/testgen.cpp
+++ b/backends/p4tools/modules/testgen/testgen.cpp
@@ -147,6 +147,7 @@ std::optional<AbstractTestList> generateTestsImpl(std::optional<std::string_view
                                                   const CompilerOptions &compilerOptions,
                                                   const TestgenOptions &testgenOptions,
                                                   bool writeTests) {
+    registerTestgenTargets();
     P4Tools::Target::init(compilerOptions.target.c_str(), compilerOptions.arch.c_str());
 
     // Set up the compilation context.
```

I have not pushed the fix yet because I am investigating why gtest does not fail.

Reply:
I think because it is a `FATAL_ERROR`. It throws an exception, but it is being caught somewhere.

Reply:
Yes, it's caught by the API, currently fixing that.

Reply:
The failure isn't caught by the tests because of a nasty side effect of static allocation. gtest initializes the target at some point for another tests and it remains initialized for the tests that should fail. 

#4286 is relevant here.

Reply:
Ah, yes. static allocation order is undefined :)